### PR TITLE
Added Azure resource management with API

### LINKED ISSUE #1214: Create any Azure resource
Create, update or delete any Azure resource using Azure REST API.
This module gives access to resources that are not supported via Ansible modules.

https://docs.ansible.com/ansible/latest/modules/azure_rm_resource_module.html#azure-rm-resource-module

### PR DESCRIPTION
**Pull request by JunqiZhang0**
Fix #1214

### CODE REVIEW

File: docs/source/azure.rst [hunk 1]
```diff
@@ -54,6 +54,39 @@ definition, the following options are available.
 
 ⚫ If you declare both public and private image, only the private will be taken
 
+azure_withApi
```
Comment (abraverm) [inline docs/source/azure.rst]: azure_api*

File: docs/source/examples/workspaces/azure/PinFile [hunk 2]
```diff
@@ -125,3 +125,18 @@ azure_vm_test:
...
+azure_api:
+  topology:
+    topology_name: azure_api
+    resource_groups:
+      - resource_group_name: "azure_api"
+        resource_group_type: "azure"
+        resource_definitions:
+          - role: "azure_api"
+            resource_group: ccit
+            url: "https://management.azure.com/subscriptions/dcc74c29-4db6-4c49-9a0f-ac0ee03fa17e/resourceGroups/ccit/providers/Microsoft.Compute/virtualMachines/test?api-version=2019-03-01"
```
Comment (abraverm) [inline docs/source/examples/workspaces/azure/PinFile]: I prefer any specific information such as our subscription will not be in the examples

File: docs/source/azure.rst (hunk 1 above)
Comment (JunqiZhang0) [reply docs/source/azure.rst]: Thanks!

File: docs/source/examples/workspaces/azure/PinFile (hunk 2 above)
Comment (JunqiZhang0) [reply docs/source/examples/workspaces/azure/PinFile]: OK will take this away

File: docs/source/azure.rst [hunk 3]
```diff
@@ -54,6 +54,39 @@ definition, the following options are available.
 
 ⚫ If you declare both public and private image, only the private will be taken
 
+azure_Api
```
Comment (abraverm) [inline docs/source/azure.rst]: lowercase for `Api` - `azure_api`

File: docs/source/azure.rst [hunk 4]
```diff
@@ -54,6 +54,39 @@ definition, the following options are available.
...
+Any Azure resources can be provisioned using this role, it supported by the Azure Api
+
+* Azure API <https://docs.microsoft.com/en-us/rest/api/compute/>`_
+* azure_vm module <https://docs.ansible.com/ansible/latest/modules/azure_rm_virtualmachine_module.html#id4>`_
+
+Topology Schema
+~~~~~~~~~~~~~~~
+
+Within Linchpin, the :term:`azure_withApi` :term:`resource_definition` has more
+options than what is shown in the examples above. For each :term:`azure_withApi`
```
Comment (abraverm) [inline docs/source/azure.rst]: `azure_withApi` -> `azure_api`

File: docs/source/azure.rst [hunk 5]
```diff
@@ -54,6 +54,39 @@ definition, the following options are available.
 
 ⚫ If you declare both public and private image, only the private will be taken
 
+azure_Api
+-------
+
+Any Azure resources can be provisioned using this role, it supported by the Azure Api
+
+* Azure API <https://docs.microsoft.com/en-us/rest/api/compute/>`_
+* azure_vm module <https://docs.ansible.com/ansible/latest/modules/azure_rm_virtualmachine_module.html#id4>`_
```
Comment (abraverm) [inline docs/source/azure.rst]: How azure_vm module is related?

File: docs/source/azure.rst (hunk 4 above)
Comment (abraverm) [inline docs/source/azure.rst]: which examples are you referring to?

File: linchpin/provision/roles/azure/tasks/provision_azure_withApi.yml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    client_id: "{{ auth_var['client_id'] | default(omit) }}"
+    tenant: "{{ auth_var['tenant'] | default(omit) }}"
+    secret: "{{ auth_var['secret'] | default(omit) }}"
+    subscription_id: "{{ auth_var['subscription_id'] | default(omit) }}"
+    resource_group: "{{res_def['resource_group']|default(omit)}}"
+    provider: compute
+    resource_type: "{{res_def['resource_type']|default('virtualmachinescalesets')}}"
+    resource_name: "{{res_def['resource_name']|default("myVmss")}}"
+    api_version: "{{res_def['api_version']|default("2019-03-01")}}"
+    body: { body }
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/provision_azure_withApi.yml]: Isn't this file is a leftover of when the role was called 'azure_withApi' and should be removed?

File: linchpin/provision/roles/azure/tasks/provision_azure_api.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+---
+- name: Use Azure Api to manage resources
+  azure_rm_resource:
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: `resource_type` and `resource_name` are missing

File: docs/source/azure.rst [hunk 8]
```diff
@@ -54,6 +54,39 @@ definition, the following options are available.
...
+
+Topology Schema
+~~~~~~~~~~~~~~~
+
+Within Linchpin, the :term:`azure_withApi` :term:`resource_definition` has more
+options than what is shown in the examples above. For each :term:`azure_withApi`
+definition, the following options are available.
+
++----------------------+------------+---------------+-----------------------+--------------------+
+| Parameter            | required   | type          | ansible value         | comments           |
```
Comment (abraverm) [inline docs/source/azure.rst]: `url` parameter is missing

File: linchpin/provision/roles/azure/tasks/provision_azure_api.yml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    client_id: "{{ auth_var['client_id'] | default(omit) }}"
+    tenant: "{{ auth_var['tenant'] | default(omit) }}"
+    secret: "{{ auth_var['secret'] | default(omit) }}"
+    subscription_id: "{{ auth_var['subscription_id'] | default(omit) }}"
+    resource_group: "{{res_def['resource_group']|default(omit)}}"
+    provider: "compute"
+    api_version: "{{res_def['api_version']|default('2019-03-01')}}"
+    url: "{{ res_def['url']|default(omit)}}"
+    body: "{{ lookup('file', '{{workspace}}/body.json') }}"
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: shouldn't this be `"{{ res_def['body'] }}"`?

File: linchpin/provision/roles/azure/tasks/provision_azure_api.yml [hunk 10]
```diff
@@ -0,0 +1,22 @@
+---
+- name: Use Azure Api to manage resources
+  azure_rm_resource:
+    client_id: "{{ auth_var['client_id'] | default(omit) }}"
+    tenant: "{{ auth_var['tenant'] | default(omit) }}"
+    secret: "{{ auth_var['secret'] | default(omit) }}"
+    subscription_id: "{{ auth_var['subscription_id'] | default(omit) }}"
+    resource_group: "{{res_def['resource_group']|default(omit)}}"
+    provider: "compute"
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: azure_rm_resource allows any resource to be created, why is it limited to "compute"?

**Comment by samvarankashyap** [comment]
@JunqiZhang0 Any update on this PR?

File: linchpin/provision/roles/azure/tasks/provision_azure_withApi.yml (hunk 6 above)
Comment (JunqiZhang0) [reply linchpin/provision/roles/azure/tasks/provision_azure_withApi.yml]: yes it is

**Comment by JunqiZhang0** [comment]
@samvarankashyap It needs an integration test and could be a little tricky for the API testing because the API will not wait, I'm working on adding a listener to the test

File: linchpin/provision/roles/azure/tasks/provision_azure_api.yml (hunk 9 above)
Comment (JunqiZhang0) [reply linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: No, that's really long

Comment (abraverm) [reply linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: Then it has to be documented, and remove the body parameter from the schema. I also would recommend allowing the user to decide the file name - what if the user will have number azure_api resources and require different files?

Comment (JunqiZhang0) [reply linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: Good point

File: linchpin/provision/roles/azure/tasks/teardown_azure_api.yml [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+- name: Use Azure Api to manage resources
+  azure_rm_resource:
+    client_id: "{{ auth_var['client_id'] | default(omit) }}"
+    tenant: "{{ auth_var['tenant'] | default(omit) }}"
+    secret: "{{ auth_var['secret'] | default(omit) }}"
+    subscription_id: "{{ auth_var['subscription_id'] | default(omit) }}"
+    api_version: "{{res_def['api_version']|default('2019-03-01')}}"
+    url: "{{ res_def['url']|default(omit)}}"
+    method: 'DELETE'
+    body: "{{ lookup('file', '{{workspace}}/body.json') }}"
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/teardown_azure_api.yml]: update this line

File: linchpin/provision/roles/azure/tasks/provision_azure_api.yml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  delay: 30
+
+- name: "Append outputitem to topology_outputs"
+  set_fact:
+    topology_outputs_azure_api: "{{ topology_outputs_azure_api + [res_def_output] }}"
+  when: res_def_output['changed'] == true
+
+- name: "Add type to resource"
+  set_fact:
+    topology_outputs_azure_api: "{{ topology_outputs_azure_api | add_res_data(lookup('vars', 'role_name'),'azure_virtual_network') }}"
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: `azure_virtual_network` - typo

File: docs/source/azure.rst [hunk 13]
```diff
@@ -54,6 +54,42 @@ definition, the following options are available.
...
 ⚫ If you declare both public and private image, only the private will be taken
 
+azure_api
+-------
+
+Any Azure resources can be provisioned using this role, it supported by the Azure Api
+
+* Example <workspaces/azure/Pinfile>`
+* azure_api module <https://docs.ansible.com/ansible/latest/modules/azure_rm_resource_module.html#azure-rm-resource-module>`_
+* Azure API <https://docs.microsoft.com/en-us/rest/api/compute/>`_
```
Comment (abraverm) [inline docs/source/azure.rst]: I think a better link would be https://docs.microsoft.com/en-us/rest/api/?view=Azure

File: linchpin/provision/roles/azure/tasks/teardown_azure_api.yml (hunk 11 above)
Comment (JunqiZhang0) [reply linchpin/provision/roles/azure/tasks/teardown_azure_api.yml]: Thank you!

File: linchpin/provision/roles/azure/tasks/provision_azure_api.yml [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+- name: Use Azure Api to manage resources
+  azure_rm_resource:
+    client_id: "{{ auth_var['client_id'] | default(omit) }}"
+    tenant: "{{ auth_var['tenant'] | default(omit) }}"
+    secret: "{{ auth_var['secret'] | default(omit) }}"
+    subscription_id: "{{ auth_var['subscription_id'] | default(omit) }}"
+    api_version: "{{res_def['api_version']|default('2019-03-01')}}"
+    url: "{{ res_def['url']|default(omit)}}"
+    body: "{{ lookup('file', body_path) }}"
```
[the remark this answers is not in the document]
Comment (JunqiZhang0) [reply linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: Will do

**Comment by JunqiZhang0** [comment]
[test]

File: config/Dockerfiles/tests.d/azure/05_azure_api [hunk 15]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash -xe
+  
+# Verify the azure vm provisioning
+# distros.exclude: none
+# providers.include: azure
+# providers.exclude: none
+
+DISTRO=${1}
+PROVIDER=${2}
+subscriptionid='dcc74c29-4db6-4c49-9a0f-ac0ee03fa17e'
```
[the remark this answers is not in the document]
Comment (JunqiZhang0) [reply config/Dockerfiles/tests.d/azure/05_azure_api]: Because of the subscription id?

**Comment by JunqiZhang0** [comment]
[test]

**Comment by JunqiZhang0** [comment]
[test]

**Comment by JunqiZhang0** [comment]
[TEST]

**Comment by JunqiZhang0** [comment]
[test]

File: config/Dockerfiles/tests.d/azure/05_azure_api (hunk 15 above)
[the remark this answers is not in the document]
Comment (abraverm) [reply config/Dockerfiles/tests.d/azure/05_azure_api]: yes, its hardcoded in the test. (line 11 too)

File: docs/source/examples/workspaces/azure/body.json [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    },
+    "osProfile": {
+      "adminUsername": "Linchpin",
+      "computerName": "myVM",
+      "adminPassword": "IloveLinchpin!"
+    },
+    "networkProfile": {
+      "networkInterfaces": [
+        {
+          "id": "/subscriptions/dcc74c29-4db6-4c49-9a0f-ac0ee03fa17e/resourceGroups/ccit/providers/Microsoft.Network/networkInterfaces/TestingNIC
```
[the remark this answers is not in the document]
Comment (abraverm) [reply docs/source/examples/workspaces/azure/body.json]: Github still shows the JSON is bad

File: linchpin/provision/roles/azure/tasks/provision_azure_api.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (abraverm) [reply linchpin/provision/roles/azure/tasks/provision_azure_api.yml]: its not documented and 'body' still there.

File: requirements.txt [hunk 17]
```diff
@@ -5,7 +5,7 @@ boto3>=1.9.96
 apache-libcloud>=0.20.1
 click>=7.0
 yamlordereddictloader>=0.4.0
-ansible>=2.7.1,<=2.9.0
+ansible>=2.7.1,<=2.8.6
```
Comment (abraverm) [inline requirements.txt]: why requirements changed part of this PR?

File: linchpin/provision/roles/azure/tasks/teardown_azure_network_interface.yml [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+    client_id: "{{ auth_var['client_id'] | default(omit) }}"
+    tenant: "{{ auth_var['tenant'] | default(omit) }}"
+    secret: "{{ auth_var['secret'] | default(omit) }}"
+    subscription_id: "{{ auth_var['subscription_id'] | default(omit) }}"
+    resource_group: "{{ res_def['resource_group'] | default(omit) }}"
+    virtual_network: "{{ res_def['virtual_network_name']|default(omit) }}"
+    subnet_name: "{{ res_def['subnet_name']|default(omit) }}"
+    name: "{{ res_def['network_interface_name']|default('network_interface_name') }}"
+    state: absent
+  register: res_def_output
```
Comment (abraverm) [inline linchpin/provision/roles/azure/tasks/teardown_azure_network_interface.yml]: Why is this part of the PR??

**Comment by JunqiZhang0** [comment]
[Test]

**Comment by JunqiZhang0** [comment]
[test]

**Comment by JunqiZhang0** [comment]
[test]

**Comment by JunqiZhang0** [comment]
[test]

**Comment by JunqiZhang0** [comment]
[test]

**Comment by JunqiZhang0** [comment]
[test]

**Comment by JunqiZhang0** [comment]
[test]

**Comment by abraverm** [comment]
@JunqiZhang0, the changes in other tests are jusr for faster feeback loop?

**Comment by JunqiZhang0** [comment]
[test]

**Comment by JunqiZhang0** [comment]
> @JunqiZhang0, the changes in other tests are jusr for faster feeback loop?

No, they just need to be tested

**Comment by JunqiZhang0** [comment]
[merge]

**Pull request merged**